File: Quick Select/347. Top K Frequent Elements/347.top-k-frequent-elements.cpp
/*
 * @lc app=leetcode id=347 lang=cpp
 *
 * [347] Top K Frequent Elements
 */

// @lc code=start
class Solution
{
public:
    vector<int> topKFrequent(vector<int> &nums, int k)
    {
        unordered_map<int, int> cnts;
        for (auto i : nums)
        {
            cnts[i]++;
        }
        vector<pair<int, int>> arr;
        for (auto &p : cnts)
        {
            arr.push_back(p);
        }

        int f = quickSelect(arr, k, 0, arr.size() - 1);

        vector<int> res;
        for (auto &p : arr) // p: {element, cnt}
        {
            if (p.second >= f)
                res.push_back(p.first);
        }
        return res;
    }

    int quickSelect(vector<pair<int, int>> &arr, int k, int begin, int end)
    {
        int pivot = begin + (end - begin) / 2;
        int p = arr[pivot].second;

        // rank(arr[p])
        int i = 0;   // i is first element after smaller
        int j = end; // j is the last element before greater
        int t = 0;   // is element being examined
        while (t <= j)
        {
            if (arr[t].second < p)
            {
                swap(arr[i], arr[t]);
                i++;
                t++;
            }
            else if (arr[t].second > p)
            {
                swap(arr[t], arr[j]);
                --j;
            }
            else
            {
                ++t;
            }
        }

        int leftSize = i - begin;
        int rightSize = end - j;

        if (rightSize >= k)
            return quickSelect(arr, k, j + 1, end);
        else if (end - i + 1 >= k)
            return p;
        else
            return quickSelect(arr, k - (end - i + 1), begin, i - 1);
    }
};

//  [SSS SPPPPGGGG]
//            i
//            t
//           j
// @lc code=end

File: Sorting/Quick Select/347. Top K Frequent Elements/347. Top K Frequent Elements.java
class Solution {
    
    List<Pair<Integer, Integer>> freq = new ArrayList<>();
    
    public int[] topKFrequent(int[] nums, int k)
    {
        Map<Integer, Integer> mp = new HashMap<>();
        for(int i: nums)
        {    
            mp.put(i, mp.getOrDefault(i, 0)+1);
        }
        
        for(int w: mp.keySet())
        {
            freq.add(new Pair<Integer, Integer>(mp.get(w), w));
        }
        int kthLargest = quickSelect(k, 0, freq.size()-1);
                
        int[] ans = new int[k];
        List<Integer> res = new ArrayList<>();
        for(Pair<Integer, Integer> p: freq){
            if(p.getKey() >= kthLargest)
            {
                res.add(p.getValue());
            }
        }
        
        
        for(int pos = 0; pos < res.size(); ++pos)
        {
            ans[pos] = res.get(pos);
        }
        return ans;
    }
    
    private int quickSelect(int k, int beg, int end)
    {    
        int pivot = freq.get(beg).getKey();
        
        // rank by pivot, like leetcode 75
        int w = beg; // checking
        int i = beg;
        int j = end;
        while(w <= j)
        {
            if(freq.get(w).getKey() < pivot)
            {
                swap(i, w);
                ++i;
                ++w;
            }
            else if(freq.get(w).getKey() > pivot)
            {
                swap(j, w);
                --j;
            }
            else
            {
                ++w;
            }
        }
        
        int leftSize = i - beg;
        int rightSize = end - j;
        
        if (rightSize>=k) 
        {
            return quickSelect(k, j+1, end);
        }
        else if (end-i+1>=k) 
        {
            return pivot;
        }
        else
        {
            return quickSelect(k-(end-i+1), beg, i-1);
        }
    }
    
    private void swap(int i, int j)
    {    
        var temp = freq.get(i);
        freq.set(i, freq.get(j));
        freq.set(j, temp);
    }
}
// beg        end
// SSSSSPPPPLLLL
//      i
//         j 
//          w 